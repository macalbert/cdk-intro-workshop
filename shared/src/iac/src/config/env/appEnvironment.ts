export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
