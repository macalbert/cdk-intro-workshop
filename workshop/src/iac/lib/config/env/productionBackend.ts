import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
