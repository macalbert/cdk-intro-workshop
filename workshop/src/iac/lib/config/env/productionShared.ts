import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
