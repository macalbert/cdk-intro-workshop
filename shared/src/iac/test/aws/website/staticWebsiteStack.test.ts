import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
	StaticWebsiteStack,
	type StaticWebsiteStackProps,
} from "../../../src/aws/website/staticWebsiteStack";
import { AppEnvironment } from "../../../src/config/env/appEnvironment";

describe("Static website Stack", () => {
	const env = {
		account: "account",
		region: "eu-west-1",
	};

	test("Should_CreateWebsite_When_StackIsCalled", () => {
		// Arrange
		const stack = new Stack(new App(), "staticWebsiteStackTest", {
			env: env,
		});
		const props = createStaticWebsiteStackProps();

		// Act
		const actual = new StaticWebsiteStack(stack, props);

		// Assert
		const template = Template.fromStack(actual).toJSON();

		for (const key in template.Resources) {
			const resource = template.Resources[key];
			if (resource.Properties.SourceObjectKeys) {
				delete resource.Properties.SourceObjectKeys;
			}
			if (resource.Properties.Content?.S3Key) {
				delete resource.Properties.Content.S3Key;
			}
			if (resource.Properties.Code?.S3Key) {
				delete resource.Properties.Code.S3Key;
			}
		}

		expect(template).toMatchSnapshot("staticWebsiteStackTest");
	});

	function createStaticWebsiteStackProps(): StaticWebsiteStackProps {
		return {
			env: env,
			name: "websitetest",
			recordName: `test`,
			domainName: "domain.com",
			distFolderPath: ".",
			certificateArn: `arn:aws:acm:us-east-1:${env.account}:certificate/some-guid`,
			hostedZoneId: "123456789",
			envName: AppEnvironment.Test,
			githubRepo: "test-website",
			stackName: `test-web-app`,
		};
	}
});
