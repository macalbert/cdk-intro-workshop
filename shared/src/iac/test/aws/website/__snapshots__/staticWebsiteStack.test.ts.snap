// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Static website Stack Should_CreateWebsite_When_StackIsCalled: staticWebsiteStackTest 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Outputs": {
    "CloudFrontDistributionDomainName": {
      "Description": "CloudFront distribution domain",
      "Export": {
        "Name": "test-website-CloudFrontDistributionDomainName",
      },
      "Value": {
        "Fn::GetAtt": [
          "distribution114A0A2A",
          "DomainName",
        ],
      },
    },
    "DnsRecordName": {
      "Description": "The DNS record name",
      "Export": {
        "Name": "test-website-AliasRecord",
      },
      "Value": {
        "Ref": "webDomainRecord45699F3C",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-account-eu-west-1",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "deploystaticwebsiteAwsCliLayer31B11195",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-account-eu-west-1",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-account-eu-west-1/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "staticwebsitebucket7EF1E06C",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "staticwebsitebucket7EF1E06C",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-account-eu-west-1",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "staticwebsitebucket7EF1E06C",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "deploystaticwebsiteAwsCliLayer31B11195": {
      "Properties": {
        "Content": {
          "S3Bucket": "cdk-hnb659fds-assets-account-eu-west-1",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "deploystaticwebsiteCustomResource7923FA82": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "staticwebsitebucket7EF1E06C",
        },
        "DistributionId": {
          "Ref": "distribution114A0A2A",
        },
        "DistributionPaths": [
          "/*",
        ],
        "OutputObjectKeys": true,
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          "cdk-hnb659fds-assets-account-eu-west-1",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "distribution114A0A2A": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "test.domain.com",
          ],
          "CustomErrorResponses": [
            {
              "ErrorCachingMinTTL": 10,
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/403.html",
            },
            {
              "ErrorCachingMinTTL": 10,
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/404.html",
            },
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "FunctionAssociations": [
              {
                "EventType": "viewer-request",
                "FunctionARN": {
                  "Fn::GetAtt": [
                    "testdomaincomurlrewrite603CDF99",
                    "FunctionARN",
                  ],
                },
              },
            ],
            "TargetOriginId": "staticWebsiteStackTestm47testwebsitewebsitetestteststackdistributionOrigin15B768DAE",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "staticwebsitebucket7EF1E06C",
                  "RegionalDomainName",
                ],
              },
              "Id": "staticWebsiteStackTestm47testwebsitewebsitetestteststackdistributionOrigin15B768DAE",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "originAccessIdentity402DAA30",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": "arn:aws:acm:us-east-1:account:certificate/some-guid",
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test-website",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-website-websitetest-test-stack",
          },
        ],
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "originAccessIdentity402DAA30": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Setup access from CloudFront to the bucket test.domain.com (read)",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "staticwebsitebucket7EF1E06C": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "Private",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "test.domain.com",
        "CorsConfiguration": {
          "CorsRules": [
            {
              "AllowedHeaders": [
                "*",
              ],
              "AllowedMethods": [
                "GET",
                "HEAD",
              ],
              "AllowedOrigins": [
                "*",
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
          {
            "Key": "aws-cdk:cr-owned:a654e0c3",
            "Value": "true",
          },
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test-website",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-website-websitetest-test-stack",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "staticwebsitebucketAutoDeleteObjectsCustomResource4931B7FC": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "staticwebsitebucketPolicy34945549",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "staticwebsitebucket7EF1E06C",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "staticwebsitebucketPolicy34945549": {
      "Properties": {
        "Bucket": {
          "Ref": "staticwebsitebucket7EF1E06C",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "staticwebsitebucket7EF1E06C",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "staticwebsitebucket7EF1E06C",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "staticwebsitebucket7EF1E06C",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "staticwebsitebucket7EF1E06C",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "originAccessIdentity402DAA30",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "staticwebsitebucket7EF1E06C",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "staticwebsitebucket7EF1E06C",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "originAccessIdentity402DAA30",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "staticwebsitebucket7EF1E06C",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testdomaincomurlrewrite603CDF99": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "
                  function handler(event) {
                    var request = event.request;
                    var uri = request.uri;
                    var staticFileExtensions = ['.html', '.js', '.css', '.png', '.jpg', '.jpeg', '.gif', '.svg', '.ico', '.woff', '.woff2', '.ttf', '.otf', '.eot', '.json', '.xml'];
                    function hasKnownExtension(uri) {
                      return staticFileExtensions.some(extension => uri.endsWith(extension));
                    }
                    // Remove trailing slash if it's not the homepage
                    if (uri.endsWith('/') && uri !== '/') {
                      uri = uri.slice(0, -1);
                    }
                    // Append .html if the URI does not represent the homepage or a static file
                    if (uri !== '/' && !hasKnownExtension(uri)) {
                      uri += '.html';
                    }
                    request.uri = uri;
                    return request;
                  }
                ",
        "FunctionConfig": {
          "Comment": "eu-west-1test-web-apptestncom-url-rewrite979D13F7",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": "eu-west-1test-web-apptestncom-url-rewrite979D13F7",
      },
      "Type": "AWS::CloudFront::Function",
    },
    "webDomainRecord45699F3C": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "distribution114A0A2A",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "123456789",
        "Name": "test.domain.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
